### PR TITLE
C++: Improve performance of `cpp/cleartext-storage-buffer`

### PR DESCRIPTION
**Pull request by MathiasVP**
This PR contains two unrelated performance fixes to prevent the `cpp/cleartext-storage-buffer` query from blowing up some strange codebases.

The first one is a general fix to global variable dataflow: We prune the relation so that we don't consider `use2(x)` to be a jump step from the global variable to the use of `x` in:
```c
int x;

void test1() {
  use1(x); // This is a jump step.
  x = 42;
  use2(x); // But this is no longer a jump step.
}
```
and similarly, for writes to global variables we only register the "final" write as a write to the global variable. So for instance, `x = 42` in:
```c
int x;

void test2() {
  x = 42; // No longer a jump step.
  x = 99; // This is still a jump step, though.
}
```

The second performance fix is to block flow through integral typed expressions in the query. We have used similar sanitizers in other queries to fix FPs, so this is a natural thing to do in this query as well, and it seems to dramatically improve the query's performance.

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll [hunk 1]
```diff
@@ -239,28 +239,81 @@ OutNode getAnOutNode(DataFlowCall call, ReturnKind kind) {
...
+}
+
+/** Holds if `i1` is always executed after `i2` */
+bindingset[i1, i2]
+private predicate instrStrictlyPostDominates(Instruction i1, Instruction i2) {
+  i1.getBlock().strictlyPostDominates(i2.getBlock())
+  or
+  exists(IRBlock block, int index1, int index2 |
+    block = i1.getBlock() and
+    block = i2.getBlock() and
```
Comment (geoffw0) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:257]: Are these two lines necessary?

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll [hunk 2]
```diff
@@ -239,28 +239,81 @@ OutNode getAnOutNode(DataFlowCall call, ReturnKind kind) {
...
+}
+
+predicate jumpInto(GlobalOrNamespaceVariable v, Node n1, Node n2) {
+  exists(StoreInstruction store |
+    store = n1.asInstruction() and
+    writesToGlobal(store, v) and
+    // Only do a jump step if there is no write that is guarenteed to
+    // overwrite the global variable.
+    not exists(StoreInstruction anotherStore |
+      store != anotherStore and
```
Comment (geoffw0) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:285]: Doesn't the "strictly" in `instrStrictlyPostDominates` already give us this?

**Review by geoffw0 — COMMENTED** [review]
I'm very glad to see some performance improvements for this query.  Two concerns:

(1) is this stuff tested somewhere?

(2) I'd like to know what happens in the case where there is a read between the two writes, or a write between the two reads?  For example:

```
int x = 0; // our global

void test()
{
  x = 10;
  printf("x = %i\n", x); // does x = 10 flow to here?
  x = 20;
}
```

In that case `x` could have been a local and maybe the "jump step" isn't needed anyway; but we can obscure the read, for example:

```
int x = 0; // our global

void useX()
{
  printf("x = %i\n", x); // does x = 10 flow to here?
}

void test()
{
  x = 10;
  useX();
  x = 20;
}
```

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll (hunk 2 above)
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:285]: Oh yes, you're totally right 👍. Will fix!

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll (hunk 1 above)
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:257]: You're right. They're implied by the next to lines. Good catch!

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll (hunk 2 above)
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll]: Fixed in 7d1bfba1bef7933450afdec7088a21386b6fe729.

**Comment by MathiasVP** [comment]
> (1) is this stuff tested somewhere?

In terms of test coverage: We already had a test with an incorrect annotation (it was actually a FP) that's fixed by the `jumpStep` change. See e4debf3edf67f5491c13f74e6745e868834edcf4 for that change to our test. I can definitely add more tests that specifically exercise this PR.

In terms of DCA: I did a DCA run over the weekend (it should be backlinked from this PR), but it looks like that run failed. I'll start a new run.

> (2) I'd like to know what happens in the case where there is a read between the two writes, or a write between the two reads? For example:
>
> ```c
> int x = 0; // our global
> 
> void test()
> {
>   x = 10;
>   printf("x = %i\n", x); // does x = 10 flow to here?
>   x = 20;
> }
> ```
> In that case x could have been a local and maybe the "jump step" isn't needed anyway; but we can obscure the read, for example:
> 
> ```c
> int x = 0; // our global
> 
> void useX()
> {
>   printf("x = %i\n", x); // does x = 10 flow to here?
> }
> 
> void test()
> {
>   x = 10;
>   useX();
>   x = 20;
> }
> ```

For the first example, we'll have:
- a normal SSA step from `x = 10` to `printf("x = %i\n", x);`.
- A jump step from `x = 20` to the global variable.
In particular, we don't have any of the following:
- A jump step from `x = 10` to `printf("x = %i\n", x);`
- A jump step from the global variable to `printf("x = %i\n", x);`

So to answer the first question in the comment: yes, `x = 10` should flow there by a local SSA step.

The second example is more sinister. I _think_ we're now losing the flow from `x = 10` to `printf("x = %i\n", x);`. I'll confirm this with a test, and think of a fix for this.

**Comment by MathiasVP** [comment]
cb56c1e262e31a47e020c02f85da233cef0c5ab9 adds your tests @geoffw0. As you correctly pointed out, the last of your examples demonstrates a problem with the pruning. I'll see if I can come up with a fix for this.

**Comment by aschackmull** [comment]
You should also add the dual test, i.e. the case when a write is reached by some call-chain in-between a local write-read sequence.
```
x = safe;
callThatMayUpdateX();
use(x); // <- ought to be a jump-step target to account for callThatMayUpdateX
```
Note that the implementations in Java and C# accounts for these sorts of updates in their field-based SSA by depending on virtual dispatch and making any call that may transitively reach a write into a may-write ssa definition.

**Comment by MathiasVP** [comment]
> You should also add the dual test, i.e. the case when a write is reached by some call-chain in-between a local write-read sequence.
> 
> ```
> x = safe;
> callThatMayUpdateX();
> use(x); // <- ought to be a jump-step target to account for callThatMayUpdateX
> ```
> 
> Note that the implementations in Java and C# accounts for these sorts of updates in their field-based SSA by depending on virtual dispatch and making any call that may transitively reach a write into a may-write ssa definition.

Hmm, yes. Good point. Thanks!

Since this turns out to be kinda difficult to do on an ad-hoc basis I may delay this change until we've merged another upcoming big change to C++ SSA.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:257]: The same change can be applied to `instrStrictlyPostDominates` as well.

**Comment by geoffw0** [comment]
Thanks for adding the tests (and explaining why my first example is not an issue).

**Pull request closed without merging**